doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+Among other APIs that interact with the host environment, configuration APIs deserve a particularly careful treatment. It is not practical to move all configuration access into activities. Moreover, .NET workflow host is initialized in a manner that closely follows the conventions for configuring an ASP.NET Core application host. That approach relies on the ASP.NET Core dependency injection mechanism for runtime component dependencies as well as for configuration. The SDK helps developers to avoid pitfalls around non-deterministic configuration data by making configuration access a first-class SDK feature: The SDK Engine plugs into the application host's dependencies provider (aka "service provider") and makes sure that any configuration data resolved for a particular workflow instance is automatically persisted using Temporal history marker-events. If a workflow is replayed, previously resolved configuration data is sourced from the recorded workflow history.
+
+### Workflow Host Configuration
+
+The user-facing foundation of the SDK is the host. It is configured in a manner that closely follows the conventions for configuring an ASP.NET Core application host ([basic example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part1_1_TypicalWorkflow.cs#L73-L89) | [more examples](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_1_BasicWorkflowUsage.cs#L67-L202) | [another example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_2_BasicWorkflowUsage_MultipleWorkers.cs#L108-L158)).
+
+Users provide workflow implementations and activity implementations in one of two ways:
+ - Pattern based: Entry points of workflow and activity implementations follow specific, well known patterns. Developers use attributes to provide metadata about such entry point to the host. The host automatically loads, and executes workflows and activities. This approach allows writing workflows (and activities) with hardly any scaffolding. 
+- Interface based: Developers implement specific interfaces to create workflows (and activities), and point the host to these implementations. The host automatically loads, and executes workflows and activities. This approach requires a little more scaffolding. However, it also allows additional functionality and helps addressing some advanced scenarios (e.g. dynamic workflow APIs).
+
+### Basic Workflow Execution
+
+Consider a sample workflow of type `SomeWorkflow` that runs a loop executing `SomeActivity` until a certain condition is met. When such condition is eventually met, the workflow is notified via a `NotifyConditionMet` signal.
+
+At application startup, the application is configured to host a worker that can execute workflows (and activities). Then, the worker is configured to poll `SomeTaskQueue` for work items, and to support execution of `SomeWorkflow` instances:
+
+```cs
+public static void Main(string[] args)
+{     
+    IHost appHost = Host.CreateDefaultBuilder(args)
```

Comment:
ASP.NET "host" is how .NET web apps are started app.
If we follow that approach, we permit dependency injection, configuration from different source, and allowing workers and exposing remote, app specific APs, all at the same time and using a de-facto standard pattern.

Doing what you suggest above is, indeed, shorter. But once we want to tackle the above-mentioned scenarios it would get longer, would it not?

Reply:
I would have to see what those scenarios look like. At its root, a worker connects to a server over gRPC and "runs" and "does things". If this "host" does a lot more things, maybe then it's the host piece that can leverage a worker instead of the other way around.

As with all the rest of this, take the simplest approach first, and then complicate it with additional utilities when you add use cases. Don't complicate the simplest approach just because of some potentially unused features.

Reply:
The host is the .NET abstraction for the thing in the app that
- reads configuration (from files of code) and provides it to components that require it,
- sets up and runs the DI container
- instantanes, starts and provides the thing that handles the requests
- etc....
In the case of Temporal "thing that handles the requests" is the worker. And the "requests" here are items from the Task queue.

From what I have heard from other folks (please correct me if I am wrong), people write apps that so more than execute workflows. For example, if you were to write an app that executes workflows and also exposes some rest APIs, then you would use the same ASP.NET host in the same way, only you would set it up with two "things that handle the requests": the Temporal worker and the HTTP web requests pipeline.

Does this clarify things?

Reply:
It clarifies a bit. The worker needs to be the smallest, lightest weight thing possible and if they want configuration files, DI, hosts, and all that mess, they can have them. We shouldn't force that on them.

> From what I have heard from other folks (please correct me if I am wrong), people write apps that so more than execute workflows. For example, if you were to write an app that executes workflows and also exposes some rest APIs, then you would use the same ASP.NET host in the same way, only you would set it up with two "things that handle the requests": the Temporal worker and the HTTP web requests pipeline.

I think it's important not to develop this with the assumption of how one may use it. They may use it as part of their host-based app or in ASP.Net or whatever. Or they may use it as some embedded mono thing at a really low level. Who knows. Make it as small and simple as possible and if there is value for providing extras on top, we can add things for that.